src/core/nativecall.c

```diff
@@ -293,19 +293,17 @@ char * MVM_nativecall_unmarshal_string(MVMThreadContext *tc, MVMObject *value, M
                 str = MVM_string_ascii_encode_malloc(tc, value_str);
                 break;
             case MVM_NATIVECALL_ARG_UTF16STR:
-                str = MVM_string_utf16_encode(tc, value_str, 0);
-#ifdef MVM_USE_MIMALLOC
-                {
-                    // strlen is wrong for UTF-16. Clearly we don't test
-                    // this. But how do we find the length?
-                    size_t str_len = strlen(str) + 1;
-                    char *libc_str = malloc(str_len);
-                    memcpy(libc_str, str, str_len);
-                    MVM_free(str);
-                    str = libc_str;
-                }
-#endif
+            {
+                MVMuint64 output_size;
+                /* This adds a (two byte) U-0000 at the end, but doesn't include
+                 * that code point in the returned output_size. */
+                char *temp = MVM_string_utf16_encode_substr(tc, value_str, &output_size, 0, -1, NULL, 0);
+                output_size += 2;
+                str = malloc(output_size);
```

Comment:
Pants again. There is actually a bug in this code (that I think you spotted and commented on, but I never acknowledged, so it got forgotten). `MVM_string_utf16_encode_substr` returns a size in whatever-the-word-is-for-16bit-units. The correct(ed) code was in the comment I made at the top level of this ticket:

            /* output_size is in 2 byte units, and doesn't include the
             * terminating U-0000, which was written. */
            size_t octet_len = 2 * (output_size + 1);
            libc_str = malloc(octet_len);
            memcpy(libc_str, str, octet_len);


So the only mistakes/problems I can spot in your branch were caused by me. Sorry.

I guess amend the two commits, which probably fits with a rebase.

(The previous rebase was useful. As I had NQP and Rakudo checked out at `master` I needed to privately merge the old branch head with master to get the version checks to work.)